### PR TITLE
ENH: Enable abstract DataSetFamily subclasses

### PR DESCRIPTION
**Pull request by quantophred**
This change enables us to define "abstract" subclasses of DataSetFamily
which implement some set of attributes, but don't represent tangible
families of DataSets, and thus don't have meaningful values for
`extra_dims`.

Previously, all subclasses of DataSetFamily had to define non-empty
`extra_dims`. Subclasses were detected in DataSetFamilyMeta by checking
if the base classes of the class under construction were specified as
anything other than the designated sentinel value: if so, `extra_dims`
was transformed to an OrderedDict and checked for non-emptiness.

With this change, the metaclass instead checks for an `_abstract`
attribute of the class under construction: if it is found and set to a
truthy value, `extra_dims` (and `_SliceType`) is left unchecked and
unmodified. Regardless of its value, the `_abstract` attribute is
removed from the constructed class' body: each subclass which wishes to
remain abstract must specify this attribute again.

### CODE REVIEW

File: zipline/pipeline/data/dataset.py [hunk 1]
```diff
@@ -609,25 +608,24 @@ def __new__(cls, name, bases, dict_):
                 columns[k] = v
                 dict_[k] = _DataSetFamilyColumn(k)
 
-        is_base_class = bases == (cls._base_marker,)
-        if is_base_class:
-            bases = (object,)
+        is_abstract = dict_.pop('__abstract__', False)
```
Comment (llllllllll) [inline zipline/pipeline/data/dataset.py]: `__{}__` is reserved for the implementation of Python. I would assume this interacts with `ABCMeta, not part of our code. We should use something like `_abstract`.

File: tests/pipeline/test_multidimensional_dataset.py [hunk 2]
```diff
@@ -60,13 +60,38 @@ class MD2(DataSetFamily):
...
     def test_empty_extra_dims(self):
-        expected_msg = (
-            'DataSetFamily must be defined with non-empty extra_dims'
+        msg = (
+            "^DataSetFamily must be defined with non-empty extra_dims,"
+            " or with `__abstract__ = True`$"
         )
-        with assert_raises_str(ValueError, expected_msg):
-            class MD(DataSetFamily):
+        with self.assertRaisesRegexp(ValueError, msg):
```
Comment (llllllllll) [inline tests/pipeline/test_multidimensional_dataset.py]: why the change to a regex with no pattern?

Comment (quantophred) [reply tests/pipeline/test_multidimensional_dataset.py]: Didn't see that we had an `assert_raises_str`. I'll use that instead. Thanks!
